src/differentiation_interface.jl

```diff
@@ -0,0 +1,72 @@
+using TaylorSeries
+
+"""
+
+    value_and_derivative(f, x)
+
+Evaluate f(x) and f'(x) at the same time. 
+We define a single method to do this because often there are shared subexpressions between a function and its derivatives,
+or they are computed together by automatic differentiation packages, so this is often more efficient than having two
+separate functions for f and f' and calling them separately.
+
+The generic implementation uses `TaylorSeries` to compute derivatives. It can be overwritten for individual functions:
+```
+f = x -> x^2
+value_and_derivative(f::typeof(f), x) = (x^2, 2x)
+```
+
+See @define_with_derivatives to define derivatives with three separate expressions.
+"""
+function value_and_derivative(f, x)
```

Comment:
Sorry, this commit ended up in the wrong branch by mistake; I am also working on refactoring derivatives but in a different branch `RefactorDerivatives`. So for now the functions in this file are not called at all anywhere else in the code. The plan is using that functions in Contractors.jl:89-90, but for now it's not there yet.

If you don't have strong opinions against it, I'll leave the commit in place because of laziness -- I don't want to look up how to cherry-pick a commit *out* of a branch in git, and this is code that will get merged anyway in the future anyway. :)

Reply:
No strong opinions, don't worry!